ast/Trees.h

```diff
@@ -1274,11 +1167,166 @@ EXPRESSION(ConstantLit) {
     //
     // We can represent all three of these in 16 bytes of space by doing a bunch of manual
     // pointer tagging.
-    const core::LocOffsets loc_;
-    ResolutionScopesOrSymbol resolutionScopesOrSymbol;
-    std::unique_ptr<UnresolvedConstantLit> original_;
+    enum class Tag : uint8_t {
+        KnownSymbol = 1,
+        ResolvedSymbol,
+        FailedResolutionSymbol,
+    };
+
+    struct Storage {
+        using tagged_storage = uint64_t;
+
+        static constexpr tagged_storage TAG_MASK = 0xffff;
+        static constexpr tagged_storage PTR_MASK = ~TAG_MASK;
+
+        // This member is a tagged quantity; its upper 48 bits will be either a raw symbol
+        // ID or a pointer to a vector of SymbolRefs representing the resolution scopes
+        // for a constant that failed to resolve.
+        //
+        // The tag determines the value for both this member and ptr2.
+        tagged_storage ptr1;
+        // This member is either a LocOffsets (in the case of a known symbol) or a pointer
+        // to the associated UnresolvedConstantLit.
+        tagged_storage ptr2;
+
+        tagged_storage untaggedPtr1() const noexcept {
+            auto val = ptr1 & PTR_MASK;
+            return val >> 16;
```

Comment:
Do we need to shift right at a signed type to ensure that we get the appropriate sign-extension behavior for the upper bits?

Reply:
Ah, this is a subtle callout.  Yes, we do; I think we mostly get away with it on x86-64 because high addresses with the relevant bit set are only used by the kernel (?).

I don't know what the requirements on arm64 are like, but it's probably a good idea there too.  I vote for doing this as a followup, because I think there's some arm64 investigation to be done, and maybe we can roll everything up into a TaggedPtrStorage<TagType> template or something.

Reply:
That sounds good; this is definitely something we should reuse across the other tagged pointer implementations.